pkg/tpm/tcpa_log.go

```diff
@@ -0,0 +1,730 @@
+		return nil, err
+	}
+
+	eventInfo := fmt.Sprintf("Blob address - 0x%d - with size - %db", platformFirmwareBlob.blobBase, platformFirmwareBlob.blobLength)
+	return &eventInfo, nil
+}
+
+func getGPTEventString(eventData []byte) (*string, error) {
+	var eventReader = bytes.NewReader(eventData)
+	var gptEvent EFIGptData
+
+	if err := binary.Read(eventReader, binary.LittleEndian, &gptEvent.uefiPartitionHeader.Signature); err != nil {
+		return nil, err
+	}
+
+	if err := binary.Read(eventReader, binary.LittleEndian, &gptEvent.uefiPartitionHeader.Revision); err != nil {
+		return nil, err
+	}
+
+	if err := binary.Read(eventReader, binary.LittleEndian, &gptEvent.uefiPartitionHeader.Size); err != nil {
+		return nil, err
+	}
+
+	if err := binary.Read(eventReader, binary.LittleEndian, &gptEvent.uefiPartitionHeader.CRC); err != nil {
+		return nil, err
+	}
+
+	if err := binary.Read(eventReader, binary.LittleEndian, &gptEvent.uefiPartitionHeader.HeaderStartLBA); err != nil {
+		return nil, err
+	}
+
+	if err := binary.Read(eventReader, binary.LittleEndian, &gptEvent.uefiPartitionHeader.HeaderCopyStartLBA); err != nil {
+		return nil, err
+	}
+
+	if err := binary.Read(eventReader, binary.LittleEndian, &gptEvent.uefiPartitionHeader.FirstUsableLBA); err != nil {
+		return nil, err
+	}
+
+	if err := binary.Read(eventReader, binary.LittleEndian, &gptEvent.uefiPartitionHeader.LastUsableLBA); err != nil {
+		return nil, err
+	}
+
+	if err := binary.Read(eventReader, binary.LittleEndian, &gptEvent.uefiPartitionHeader.DiskGUID); err != nil {
+		return nil, err
+	}
+
+	// Stop here we only want to know which device was used here.
+
+	eventInfo := fmt.Sprint("Disk Guid - ")
+	eventInfo += gptEvent.uefiPartitionHeader.DiskGUID.String()
+	return &eventInfo, nil
+}
+
+func getImageLoadEventString(eventData []byte) (*string, error) {
+	var eventReader = bytes.NewReader(eventData)
+	var imageLoadEvent EFIImageLoadEvent
+
+	if err := binary.Read(eventReader, binary.LittleEndian, &imageLoadEvent.imageLocationInMemory); err != nil {
+		return nil, err
+	}
+
+	if err := binary.Read(eventReader, binary.LittleEndian, &imageLoadEvent.imageLengthInMemory); err != nil {
+		return nil, err
+	}
+
+	if err := binary.Read(eventReader, binary.LittleEndian, &imageLoadEvent.imageLinkTimeAddress); err != nil {
+		return nil, err
+	}
+
+	if err := binary.Read(eventReader, binary.LittleEndian, &imageLoadEvent.lengthOfDevicePath); err != nil {
+		return nil, err
+	}
+
+	// Stop here we only want to know which device was used here.
+
+	eventInfo := fmt.Sprintf("Image loaded at address 0x%d ", imageLoadEvent.imageLocationInMemory)
+	eventInfo += fmt.Sprintf("with %db", imageLoadEvent.imageLengthInMemory)
+
+	return &eventInfo, nil
+}
+
+func getVariableDataString(eventData []byte) (*string, error) {
+	var eventReader = bytes.NewReader(eventData)
+	var variableData EFIVariableData
+
+	if err := binary.Read(eventReader, binary.LittleEndian, &variableData.variableName.blockA); err != nil {
```

Comment:
This smells like EFIGUID should have a method ReadFrom(), if it really has to be all 5 of these, but also

this doesn't have to be 5 different reads?

```go
if err := binary.Read(eventReader, binary.LittleEndian, &variableData.variableName); err != nil {
  return nil, err
}
```

should work - it'll read them in the same order that they are defined in the struct.

from the docs: https://golang.org/pkg/encoding/binary/#Read 

> Data must be a pointer to a fixed-size value or a slice of fixed-size values.

`EFIGUID` is a fixed-size value, since it is composed of fixed-size values.